yt/yt/client/logging/dynamic_table_log_writer.cpp

```diff
@@ -0,0 +1,372 @@
+    const TLogger Logger;
+
+    DECLARE_THREAD_AFFINITY_SLOT(LoggingThread);
+
+    //! Protects the fields below.
+    YT_DECLARE_SPIN_LOCK(NThreading::TSpinLock, SpinLock_);
+    TString CurrentBuffer_;
+    i64 CurrentRowCount_ = 0;
+    //! This queue should typically contain no more than one element.
+    //! We use it to limit the number of rows written within a single transaction.
+    std::deque<TString> BufferQueue_;
+
+    //! Only accessed in the logging thread.
+    bool Suspended_ = false;
+
+    //! Stored in terms of YSON bytes.
+    std::atomic<i64> BacklogWeight_{0};
+
+    void RotateBuffer()
+    {
+        VERIFY_SPINLOCK_AFFINITY(SpinLock_);
+
+        if (CurrentBuffer_.empty()) {
+            return;
+        }
+
+        BufferQueue_.push_back(std::move(CurrentBuffer_));
+        // Moved-from state should be fine, but just to be sure.
+        CurrentBuffer_.clear();
+        CurrentRowCount_ = 0;
+    }
+
+    // Periodic executor guarantees that at most one callback is executed at once.
+    // This is important to avoid log reordering.
+    void DoFlush()
+    {
+        VERIFY_INVOKER_AFFINITY(Invoker_);
+
+        TString ysonRows;
+
+        {
+            auto guard = Guard(SpinLock_);
+
+            // Let's not increase the size of the queue if there is already something to flush.
+            if (BufferQueue_.empty()) {
+                RotateBuffer();
+            }
+
+            if (BufferQueue_.empty()) {
+                return;
+            }
+
+            ysonRows = std::move(BufferQueue_.front());
+            BufferQueue_.pop_front();
+        }
+
+        YT_VERIFY(!ysonRows.empty());
+
+        auto decrementBacklogWeight = Finally([this, size = std::ssize(ysonRows)] {
+            BacklogWeight_.fetch_sub(size, std::memory_order::relaxed);
+        });
+
+        TBackoffStrategy backoff(Config_->WriteBackoff);
+        while (true) {
+            try {
+                auto dataWeightWritten = DoFlushIteration(ysonRows);
+                IncrementSegmentSize(dataWeightWritten);
+                return;
+            } catch (const std::exception& ex) {
+                YT_LOG_ERROR(ex, "Error flushing log events to dynamic table");
+            }
+
+            if (backoff.Next()) {
+                TDelayedExecutor::WaitForDuration(backoff.GetBackoff());
+            } else {
+                YT_LOG_ERROR("Flush retries exhausted, dropping log events (Size: %v)", ysonRows.size());
+                break;
+            }
+        }
+    }
+
+    i64 DoFlushIteration(const TString& ysonRows)
+    {
+        VERIFY_INVOKER_AFFINITY(Invoker_);
+
+        auto client = ClientHolder_->Client.Load();
+
+        if (!client) {
+            THROW_ERROR_EXCEPTION("Cannot flush rows without configured YT client");
+        }
+
+        auto tableMountCache = client->GetTableMountCache();
+        auto tableInfo = WaitFor(tableMountCache->GetTableInfo(Config_->TablePath))
```

Comment:
WaitForFast can be done here to avoid a context switch in a typical case. It does not make much sense in this case, but is a good pattern anyway